week-3/mandatory/2-api/server.js

```diff
@@ -0,0 +1,307 @@
+
+//   let query = `SELECT products.product_name, pa.unit_price price, sup.supplier_name
+//   FROM products JOIN product_availability pa ON (products.id=pa.prod_id)
+//   JOIN suppliers sup ON (pa.supp_id=sup.id)`;
+
+//   let param = [];
+
+//   if (name !== undefined) {
+//     query = `SELECT products.product_name, pa.unit_price price, sup.supplier_name
+//   FROM products JOIN product_availability pa ON (products.id=pa.prod_id)
+//   JOIN suppliers sup ON (pa.supp_id=sup.id) WHERE LOWER(products.product_name) LIKE '%'||$1||'%'`;
+
+//     param = [name.toLowerCase()];
+//   }
+
+//   db.query(query, param)
+//     .then((result) => res.json(result.rows))
+//     .catch((e) => console.log(e));
+// });
+
+// ------ Create New Cust -----------
+
+app.post("/customers", (req, res) => {
+  const custName = req.body.name;
+  const custAddress = req.body.address;
+  const custCity = req.body.city;
+  const custCountry = req.body.country;
+  db.query(
+    "INSERT INTO customers (name, address, city, country) " +
+      "VALUES ($1, $2, $3, $4) " +
+      "RETURNING id",
+    [custName, custAddress, custCity, custCountry],
+    (err) => {
+      if (err === undefined) {
+        res.send("New customer added");
+      } else {
+        console.log(err);
+        res.status(500).json({ error: err });
+      }
+    }
+  );
+});
+
+// --------- Create New Product ---------
+
+app.post("/products", (req, res) => {
+  const productName = req.body.product_name;
+
+  db.query(
+    "INSERT INTO products (product_name) VALUES ($1) RETURNING id",
+    [productName],
+    (err) => {
+      if (!err) res.send("New Producted Added");
+      res.status(500).json({ error: err });
+      console.log(err);
+    }
+  );
+});
+
+// !! ----------Create New Availability -----
+
+app.post("/availability", (req, res) => {
+  const suppId = req.body.supp_id;
+  const price = req.body.unit_price;
+
+  if (price <= 0 || !Number.isInteger(price)) {
+    res.status(400).send("Please enter positive integer value for the price");
+  } else {
+    db.query(
+      "SELECT prod_id, supp_id FROM product_availability WHERE supp_id=$1",
+      [suppId],
+      (err, result) => {
+        if (!err) {
+          if (result.rowCount < 2) res.send("You cannot do that!");
+        }
+      }
+    );
+
+    db.query(
+      `INSERT INTO product_availability (supp_id, unit_price) VALUES ($1, $2)`,
+      [suppId, price],
+      (err) => {
+        if (!err) res.send("New Product Availabity Added");
+      }
+    );
+  }
+});
+
+//--------- Create a new order for a customer ----
+
+app.post("/customers/:customerId/orders", (req, res) => {
+  const customerId = parseInt(req.params.customerId);
+  const orderDate = req.body.order_date;
+  const orderRef = req.body.order_reference;
+
+  db.query(
+    "SELECT * FROM customers WHERE id = $1",
+    [customerId],
+    (error, result) => {
+      if (!error) {
+        if (result.rowCount == 0) res.send("No customer with this ID");
```

Comment:
so if there is an error it will actually try to insert it, you need to add catching for errors